README.md
# sw-proxy

sw-proxy is a really tiny proxy server which utilizes a service worker.
A [service worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API)
sits between web applications and the network (when available).

Service workers are able to intercept requests and take an action, like
serving a custom response.

This enables you to use sw-proxy as a proxy server inside your browser or as a mock server to test a REST API without 
having to run a local backend.


## Installation
Run `npm install @dannymoerkerke/sw-proxy --save-dev`

## Usage
Simply add sw-proxy to your web app using a script tag:

`<script src="node_modules/sw-proxy/index.js></script>`

Then list the requests you want to intercept in an array of objects and save it to `sw-proxy-responses.js` in the
root folder of your application.

For example:

```javascript
const responses = [
    {
         url: 'http://api.example.com/json',
         headers: {
             'Content-Type': 'application/json'
         },
         body: {
             message: 'a json response'
         },
         status: 200,
         statusText: 'OK'
    },
    {
        ...
    }
]
```


- `url`: String (required), fully qualified URL of the request to be proxied
- `method`: String (required), HTTP method (GET, POST, PUT or DELETE)
- `headers`: Object (optional), key/value pairs of HTTP headers
- `body`: Object/String (optional): response body
- `file`: String (optional), path to file to be served, relative to root folder of the
application. Must be accessible to the web server and will be ignored if `body` is present
- `status`: Number (optional), HTTP status code
- `statusText`: String (optional), HTTP status text

## Running the demo
Run `npm install`.

Rename `sw-proxy-responses.example.js` to `sw-proxy-responses.js`.

Run `npm start`.

The demo can now be viewed on http://localhost:8080/demo.

Click the buttons on the demo page to generate the requests and
view the results in the dev tools of your browser of choice.

To view the console logs in Safari you need to open a separate Web Inspector
window via Develop > Service Workers > localhost.

## Browser support
Tested in:
- Chrome 67+
- Safari 11.1+
- Firefox 60+
- Edge 17+

sw-proxy should run in all browsers that support Service Worker.

## Inspecting and debugging the service worker
For a great explanation of how to inspect and debug service workers
in Chrome, Firefox, Safari and Edge check [Tools for PWA developers](https://developers.google.com/web/ilt/pwa/tools-for-pwa-developers#interact_with_service_workers_in_the_browser).


CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.2] - 2019-18-11
### Added
- repo url

## [0.0.1] - 2019-18-11
- First version


sw-proxy.js
self.importScripts('./sw-proxy-responses.js');

const handleInstall = () => {
  console.log('[SW-PROXY] service worker installed');
  self.skipWaiting();
};

const handleActivate = () => {
  console.log('[SW-PROXY] service worker activated');

  return self.clients.claim();
};

const delayResponse = (time, response) => new Promise(resolve => setTimeout(() => resolve(response), time));
const compose = (...fns) => x => fns.reduce((res, f) => res || f(x), false);

const getResponseFor = (url, method) => {
  const exactUrlMatch = ({reqUrl, reqMethod}) => url === reqUrl && method === reqMethod;
  const patternUrlMatch = ({reqUrl, reqMethod}) => reqUrl.includes('*') && new RegExp(reqUrl.replace('*', '(.+)')).test(url) && method === reqMethod;
  const exactOrPatternMatch = compose(exactUrlMatch, patternUrlMatch);

  /* eslint-disable no-undef */
  return responses.find(exactOrPatternMatch);
  /* eslint-enable no-undef */
};

const handleFetch = async (e) => {
  const {request} = e;
  const {method: reqMethod, url: reqUrl} = request;
  const response = getResponseFor(reqUrl, reqMethod);

  if(response) {
    const {headers, status, statusText, delay, resMethod} = response;
    const redirectUrl = response.reqUrl.includes('*') ? reqUrl.replace(new RegExp(response.reqUrl.replace('*', '(.+)')), response.redirectUrl) : response.redirectUrl;
    const init = {headers, status, statusText, url: reqUrl, method: resMethod ? resMethod : reqMethod};

    const proxyResponse = response.file ? fetch(`${self.origin}/${response.file}`) :
      redirectUrl ? fetch(redirectUrl, init) :
        Promise.resolve(new Response(JSON.stringify(response.body), init));


    const msg = `[SW-PROXY] proxying request ${reqMethod}: ${reqUrl}`;
    console.log(`${msg} ${redirectUrl ? `-> ${redirectUrl}` : ``} ${response.file ? `-> serving: ${response.file}` : ``}`);

    e.waitUntil(
      Promise.resolve(e.respondWith(delay ? delayResponse(delay, proxyResponse) : proxyResponse))
    );
  }
};

self.addEventListener('install', handleInstall);
self.addEventListener('activate', handleActivate);
self.addEventListener('fetch', handleFetch);


sw-proxy-responses.example.js
/* eslint-disable no-unused-vars */
const responses = [
  {
    reqUrl: 'https://api.example.com/json',
    reqMethod: 'GET',
    headers: {
      'Content-Type': 'application/json'
    },
    body: {
      message: 'a json response'
    }
  },
  {
    reqUrl: 'https://api.example.com/text',
    reqMethod: 'GET',
    body: 'plain text body'
  },
  {
    reqUrl: 'https://api.example.com/pdf',
    reqMethod: 'GET',
    headers: {
      'Content-Type': 'application/pdf'
    },
    file: 'demo/example.pdf'
  },
  {
    reqUrl: 'https://api.example.com/post',
    reqMethod: 'POST',
    status: 201,
    statusText: 'Created',
    headers: {
      'Content-Type': 'application/json'
    },
    body: {
      message: 'created'
    }
  },
  {
    reqUrl: 'https://api.example.com/notfound',
    reqMethod: 'GET',
    status: 404,
    statusText: 'Not found',
    headers: {
      'Content-Type': 'application/json'
    },
    body: {
      message: 'Not found'
    }
  },
  {
    reqUrl: 'https://localhost/*',
    reqMethod: 'GET',
    status: 302,
    redirectUrl: 'http://localhost:8080/$1'
  }
];


sw-proxy-responses.js
/* eslint-disable no-unused-vars */
const responses = [
  {
    reqUrl: 'http://api.example.com/json',
    reqMethod: 'GET',
    headers: {
      'Content-Type': 'application/json'
    },
    body: {
      message: 'a json response'
    }
  },
  {
    reqUrl: 'http://api.example.com/text',
    reqMethod: 'GET',
    body: 'plain text body'
  },
  {
    reqUrl: 'http://api.example.com/pdf',
    reqMethod: 'GET',
    headers: {
      'Content-Type': 'application/pdf'
    },
    file: 'demo/example.pdf'
  },
  {
    reqUrl: 'http://api.example.com/post',
    reqMethod: 'POST',
    status: 201,
    statusText: 'Created',
    headers: {
      'Content-Type': 'application/json'
    },
    body: {
      message: 'created'
    }
  },
  {
    reqUrl: 'http://api.example.com/notfound',
    reqMethod: 'GET',
    status: 404,
    statusText: 'Not found',
    headers: {
      'Content-Type': 'application/json'
    },
    body: {
      message: 'Not found'
    }
  },
  {
    reqUrl: 'http://localhost:8080/redirect',
    reqMethod: 'GET',
    status: 302,
    redirectUrl: 'https://api.dannymoerkerke.com/v1/blogpostings'
  },
  {
    reqUrl: 'http://localhost/*',
    reqMethod: 'GET',
    status: 302,
    redirectUrl: 'http://localhost:8080/$1'
  }
];
